logging/context.go

```diff
@@ -0,0 +1,78 @@
+package logging
+
+import (
+	"context"
+
+	"go.viam.com/utils"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+type debugLogKeyType int
+
+const debugLogKeyID = debugLogKeyType(iota)
+
+// EnableDebugMode returns a new context with debug logging state attached. An empty `debugLogKey`
+// generates a random value.
+func EnableDebugMode(ctx context.Context, debugLogKey string) context.Context {
+	if debugLogKey == "" {
+		debugLogKey = utils.RandomAlphaString(6)
```

Comment:
why does this have to be a random string?

Reply:
Not sure if the question is about assigning a string at all, or why a random over a constant.
1) Calling `EnableDebugMode(ctx, "")` would have an unexpected result. So we should assign something.
2) I've deviated from some of the scope, but [this sub-bullet point](https://docs.google.com/document/d/1AHt9WvkjOcndY_fypw6WnKbGh3sBholbwr9n95ADAOw/edit#bookmark=id.opeoe7s356yp) hopefully gives context for what this string is for.

The idea is if a client writes code like this:
```
base.MoveStraight(EnableDebugMode(ctx, "my moveStraight op"), 10 meters, 1m/s)
```
A set of (hypothetical) debug log lines would be generated on the robot (even if the wheeled base/motors were configured for only INFO logs) like:
```
2023-12-04T10:45:08.974-0500	DEBUG	robot_server.rdk:component:base/Base "my moveStraight op"	wheeled/wheeled_base.go:387	received a SetPower with linear.X: 0.00, linear.Y: 0.50 linear.Z: 0.00, angular.X: 0.00, angular.Y: 0.00, angular.Z: 0.00
2023-12-04T10:45:08.974-0500	DEBUG	robot_server.rdk:component:motor/Left	"my moveStraight op" fake/motor.go:189	Motor SetPower 0.500000
2023-12-04T10:45:09.784-0500	DEBUG	robot_server.rdk:component:motor/Left	"my moveStraight op" fake/motor.go:354	Motor Stopped
2023-12-04T10:45:10.280-0500	DEBUG	robot_server.rdk:component:base/Base	"my moveStraight op" wheeled/wheeled_base.go:387	received a SetPower with linear.X: 0.00, linear.Y: -0.50 linear.Z: 0.00, angular.X: 0.00, angular.Y: 0.00, angular.Z: 0.00
2023-12-04T10:45:10.280-0500	DEBUG	robot_server.rdk:component:motor/Left	"my moveStraight op" fake/motor.go:189	Motor SetPower -0.500000
2023-12-04T10:45:10.684-0500	DEBUG	robot_server.rdk:component:motor/Left	"my moveStraight op" fake/motor.go:354	Motor Stopped
2023-12-04T10:45:10.934-0500	DEBUG	robot_server.rdk:component:motor/Left	"my moveStraight op" fake/motor.go:354	Motor Stopped
```

Using a random string means client code like:
```
base.MoveStraight(EnableDebugMode(ctx, ""), 10 meters, 1m/s)
base.MoveStraight(EnableDebugMode(ctx, ""), 10 meters, 1m/s)
```

Would give logs like (omitting the motors for brevity):
```
2023-12-04T10:45:08.974-0500	DEBUG	robot_server.rdk:component:base/Base "abcdef"	wheeled/wheeled_base.go:387	received a SetPower with linear.X: 0.00, linear.Y: 0.50 linear.Z: 0.00, angular.X: 0.00, angular.Y: 0.00, angular.Z: 0.00
2023-12-04T10:45:10.280-0500	DEBUG	robot_server.rdk:component:base/Base	"abcdef" wheeled/wheeled_base.go:387	received a SetPower with linear.X: 0.00, linear.Y: -0.50 linear.Z: 0.00, angular.X: 0.00, angular.Y: 0.00, angular.Z: 0.00
2023-12-04T10:45:08.974-0500	DEBUG	robot_server.rdk:component:base/Base "123456"	wheeled/wheeled_base.go:387	received a SetPower with linear.X: 0.00, linear.Y: 0.50 linear.Z: 0.00, angular.X: 0.00, angular.Y: 0.00, angular.Z: 0.00
2023-12-04T10:45:10.280-0500	DEBUG	robot_server.rdk:component:base/Base	"123456" wheeled/wheeled_base.go:387	received a SetPower with linear.X: 0.00, linear.Y: -0.50 linear.Z: 0.00, angular.X: 0.00, angular.Y: 0.00, angular.Z: 0.00
```

Reply:
Ah I was specifically wondering about the randomness. Got it, we just want something unique-ish for tracing.

Reply:
I'll leave a comment. It was hasty of me to omit one. The only reasonable response to seeing that code is the process you just went through.

Reply:
[minor] `EnableDebugMode(ctx, "")` feels a little inconvenient to me. Maybe we can make this two functions instead: `EnableDebugMode(ctx)` that sets a random value, and `EnableDebugModeWithKey(ctx, key)` that behaves like the current function.

Reply:
Done